base_layer/wallet_ffi/src/lib.rs

```diff
@@ -218,20 +230,175 @@ pub struct TariWallet {
     shutdown: Shutdown,
 }
 
-#[allow(unused)]
+#[derive(Debug)]
+#[repr(C)]
+pub enum TariUtxoSort {
+    ValueAsc = 0,
+    ValueDesc = 1,
+    MinedHeightAsc = 2,
+    MinedHeightDesc = 3,
+}
+
+#[derive(Debug, Clone)]
+#[repr(C)]
+pub struct TariUtxo {
+    pub commitment: *mut c_char,
+    pub value: u64,
+}
+
+impl TryFrom<DbUnblindedOutput> for TariUtxo {
+    type Error = InterfaceError;
+
+    fn try_from(x: DbUnblindedOutput) -> Result<Self, Self::Error> {
+        Ok(Self {
+            commitment: CString::new(x.commitment.to_hex())
+                .map_err(|e| {
+                    InterfaceError::InvalidArgument(format!("failed to obtain hex from a commitment: {:?}", e))
+                })?
+                .into_raw(),
+            value: x.unblinded_output.value.as_u64(),
+        })
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+#[repr(C)]
+pub enum TariTypeTag {
+    String = 0,
+    Utxo = 1,
+    PublicKey = 2,
+}
+
+impl Display for TariTypeTag {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            TariTypeTag::String => write!(f, "String"),
+            TariTypeTag::Utxo => write!(f, "Utxo"),
+            TariTypeTag::PublicKey => write!(f, "PublicKey"),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+#[repr(C)]
+pub struct TariVector {
+    pub tag: TariTypeTag,
+    pub len: usize,
+    pub cap: usize,
+    pub ptr: *mut c_void,
+}
+
+impl TariVector {
+    fn from_string_vec(v: &mut Vec<String>) -> Result<Self, InterfaceError> {
+        Ok(Self {
+            tag: TariTypeTag::String,
+            len: v.len(),
+            cap: v.capacity(),
+            ptr: v.as_mut_ptr() as *mut c_void,
+        })
+    }
+
+    fn from_pubkey_vec(v: &mut Vec<PublicKey>) -> Result<Self, InterfaceError> {
+        Ok(Self {
+            tag: TariTypeTag::PublicKey,
+            len: v.len(),
+            cap: v.capacity(),
+            ptr: v.as_mut_ptr() as *mut c_void,
+        })
+    }
+
+    fn to_string_vec(&self) -> Result<Vec<String>, InterfaceError> {
```

Comment:
The `to_*` functions should be marked as unsafe because they do not provide any additional safety guarantees than from_raw_parts (which is unsafe). If the data that TariVector points to is null/deallocated the Vec<String> will be invalid and the program will crash. Moreover if the vector is mutated (as you are free to do with an owned Vec, it could be reallocated and the pointer is no longer valid.

I think the method signature should be:
`unsafe fn as_string_slice(&self) -> Result<&[&str], InterfaceError>`